lib/sass_builder.dart

```diff
@@ -1,94 +1,148 @@
-            _log.info('adding mainInputFile: ${mainInputAssetId}');
-            _mainInputs.add(mainInputAssetId);
-          } else {
-            rewriteInputs = true;
-          }
-        }
-        if (rewriteInputs) {
-          _log.info('rewriting inputs');
-          _saveMainInputs();
-        }
-      } else {
-        _mainInputs = new Set();
-      }
+  SassBuilder({BarbackSettings settings}) {
+    if (settings != null) {
+      _outputExtension = settings.configuration.containsKey(outputExtensionKey)
+          ? settings.configuration[outputExtensionKey]
+          : '.css';
+    } else {
+      _outputExtension = '.css';
     }
-    return _mainInputs;
   }
 
-  _saveMainInputs() {
-    var sassBuildFile = new File(join(Directory.current.path, '.dart_tool', 'sass_build_main_inputs.json'));
-    if (!sassBuildFile.existsSync()) sassBuildFile.createSync(recursive: true);
+  @override
+  Future build(BuildStep buildStep) async {
+    var inputId = buildStep.inputId;
+
+    if (basename(inputId.path).startsWith('_')) {
+      // Do not produce any output for .scss partials.
+      _log.fine('skipping parial file: ${inputId}');
+      return;
+    }
+
+    // Read and copy this asset and all imported assets to the temp directory.
+    _log.fine('processing file: ${inputId}');
+    var tempDir =
+        await buildStep.fetchResource<ScratchSpace>(_scratchSpaceResource);
+    await _readAndCopyImports(inputId, buildStep, tempDir);
 
-    _log.info('writing mainInputs: $_mainInputs');
-    sassBuildFile.writeAsStringSync(JSON.encode(_mainInputs.map((mi) => mi.toString()).toList()));
+    // Compile the css.
+    var tempAssetPath = tempDir.fileFor(inputId).path;
+    _log.fine('compiling file: ${tempAssetPath}');
+    var cssOutput = compile(tempAssetPath,
+        packageResolver: new SyncPackageResolver.root(tempDir.packagesDir.uri));
+
+    // Write the builder output
+    var outputId = inputId.changeExtension(_outputExtension);
+    buildStep.writeAsString(outputId, '${cssOutput}\n');
+    _log.fine('wrote css file: ${outputId.path}');
   }
 
   @override
-  Future build(BuildStep buildStep) async {
-    var inputId = buildStep.inputId;
-    _log.info('processing changed file: ${inputId}');
+  Map<String, List<String>> get buildExtensions => {
+        '.scss': [_outputExtension],
+        '.sass': [_outputExtension],
+      };
 
-    if (!basename(inputId.path).startsWith('_')) {
-      if (!mainInputs.contains(inputId)) {
-        mainInputs.add(inputId);
-        _saveMainInputs();
-      }
+  // Reads `id` and all transitive imports while copying them to `tempDir`.
+  //
+  // Uses `buildStep to read the assets and the `ScratchSpace` API to write
+  // `tempDir`.
+  Future _readAndCopyImports(
+      AssetId id, BuildStep buildStep, ScratchSpace tempDir) async {
+    var assetsToCopy = new Queue<AssetId>();
+    assetsToCopy.add(id);
+
+    while (assetsToCopy.isNotEmpty) {
+      id = assetsToCopy.removeFirst();
```

Comment:
I think  you also need a `Set<AssetId>` of previously processed assets, and you could either `continue` here if `id` is already in that or make sure you don't add ones you have already seen on line 93 below.

Reply:
I had that check in my original version as a guard against copying a file twice but I removed it since the ScratchSpace now provides the guarantee for me. 

This could get stuck running forever if you gave it sass files with a cycle of imports. I'll add the check back in.